extension/storage/filestorage/client.go

```diff
@@ -117,40 +142,87 @@ func (c *fileStorageClient) Close(_ context.Context) error {
 }
 
 // Compact database. Use temporary file as helper as we cannot replace database in-place
-func (c *fileStorageClient) Compact(ctx context.Context, compactionDirectory string, timeout time.Duration, maxTransactionSize int64) (*fileStorageClient, error) {
+func (c *fileStorageClient) Compact(ctx context.Context, compactionDirectory string, timeout time.Duration, maxTransactionSize int64) error {
+	var err error
+	var file *os.File
+	var compactedDb *bbolt.DB
+
 	// create temporary file in compactionDirectory
-	file, err := ioutil.TempFile(compactionDirectory, "tempdb")
+	file, err = ioutil.TempFile(compactionDirectory, "tempdb")
 	if err != nil {
-		return nil, err
+		return err
 	}
 
+	defer func() {
+		_, statErr := file.Stat()
+		if statErr == nil {
+			// File still exists and needs to be removed
+			if removeErr := os.Remove(file.Name()); removeErr != nil {
+				c.logger.Error("removing temporary compaction file failed", zap.Error(removeErr))
+			}
+		}
+	}()
+
 	// use temporary file as compaction target
 	options := bboltOptions(timeout)
 
+	c.logger.Debug("starting compaction",
+		zap.String(directoryKey, c.db.Path()),
+		zap.String(tempDirectoryKey, file.Name()))
+
 	// cannot reuse newClient as db shouldn't contain any bucket
-	db, err := bbolt.Open(file.Name(), 0600, options)
+	compactedDb, err = bbolt.Open(file.Name(), 0600, options)
 	if err != nil {
-		return nil, err
+		return err
 	}
 
-	if err := bbolt.Compact(db, c.db, maxTransactionSize); err != nil {
-		return nil, err
+	c.compactionMutex.Lock()
+	defer c.compactionMutex.Unlock()
+	compactionStart := time.Now()
+
+	if err = bbolt.Compact(compactedDb, c.db, maxTransactionSize); err != nil {
+		return err
 	}
 
 	dbPath := c.db.Path()
-	tempDBPath := db.Path()
+	compactedDbPath := compactedDb.Path()
 
-	db.Close()
-	c.Close(ctx)
+	c.db.Close()
+	compactedDb.Close()
 
 	// replace current db file with compacted db file
-	if err := os.Remove(dbPath); err != nil {
-		return nil, err
+	if err = os.Remove(dbPath); err != nil {
+		return err
 	}
 
-	if err := os.Rename(tempDBPath, dbPath); err != nil {
-		return nil, err
+	if err = os.Rename(compactedDbPath, dbPath); err != nil {
+		return err
+	}
+
+	c.db, err = bbolt.Open(dbPath, 0600, options)
+
+	c.logger.Info("finished compaction",
+		zap.String(directoryKey, dbPath),
+		zap.Duration(elapsedKey, time.Since(compactionStart)))
+
+	return err
+}
+
+// shouldCompactOnWrite checks whether the conditions for online compaction are met
+func (c *fileStorageClient) shouldCompactOnWrite() bool {
+	if !c.compactionCfg.OnRebound {
+		return false
 	}
 
-	return newClient(dbPath, timeout)
+	dbStats := c.db.Stats()
+	if dbStats.FreelistInuse > int(c.compactionCfg.ReboundSizeBelowMiB) ||
+		dbStats.FreeAlloc+dbStats.FreelistInuse < int(c.compactionCfg.ReboundTotalSizeAboveMiB) {
+		return false
```

Comment:
There are two problems with this check.

One is that our configuration values are in MiB, but the stats values are in bytes. We need to normalize them.

The second is that we're not checking the right thing. `FreelistInuse` is just the size of the freelist data structure, and since we don't sync that to disk, it's irrelevant here. `FreeAlloc` is actually the total size of free pages, which is what we want, but I couldn't figure out how to get the size of the used pages from stats. Best I could come up with was the following:

```go
func getDbSize(db *bbolt.DB) (totalSize, dataSize int64, err error) {
	var totalSize int64

	err := c.db.View(func(tx *bbolt.Tx) error {
		totalSize = tx.Size()
		return nil
	})
	if err != nil {
		c.logger.Error("failed to get db size", zap.Error(err))
		return 0, 0, nil
	}

	dbStats := c.db.Stats()
	dataSize := totalSize - int64(dbStats.FreeAlloc)
        return totalSize, dataSize, nil
}
```

Reply:
I run a quick check to see how the two compare:

```
022-06-02T13:40:32.158+0200	DEBUG	filestorage/client.go:261	shouldCompact check	{"FreeAlloc": 21024768, "FreelistInuse": 41080}
2022-06-02T13:40:32.159+0200	DEBUG	filestorage/client.go:265	shouldCompact check	{"totalSize": 21049344, "dataSize": 24576}
2022-06-02T13:40:33.158+0200	DEBUG	filestorage/client.go:261	shouldCompact check	{"FreeAlloc": 25022464, "FreelistInuse": 48888}
2022-06-02T13:40:33.158+0200	DEBUG	filestorage/client.go:265	shouldCompact check	{"totalSize": 25051136, "dataSize": 28672}
2022-06-02T13:40:34.159+0200	DEBUG	filestorage/client.go:261	shouldCompact check	{"FreeAlloc": 25022464, "FreelistInuse": 48888}
2022-06-02T13:40:34.159+0200	DEBUG	filestorage/client.go:265	shouldCompact check	{"totalSize": 25051136, "dataSize": 28672}
2022-06-02T13:40:35.158+0200	DEBUG	filestorage/client.go:261	shouldCompact check	{"FreeAlloc": 25022464, "FreelistInuse": 48888}
2022-06-02T13:40:35.158+0200	DEBUG	filestorage/client.go:265	shouldCompact check	{"totalSize": 25051136, "dataSize": 28672}
```